Masternode Getting Started Guide (#112)
tomochain/docs: docs/get-started/masternode-setup-guide.md

Spin up a machine in your favorite cloud and get your masternode running in a few minutes!”

Install and update the tomochain-created `tmn` utility from pip:

```shell
pip3 install --user tmn
pip3 install -U tmn
```

> Watch out for WARNINGs or ERRORs and troubleshoot.

Check that `tmn` has been correctly installed, use the following command to show some tmn info:

```shell
pip3 show tmn

Name: tmn
Version: 0.2.5
Summary: Quickstart your masternode
Home-page: https://tomochain.com
Author: Etienne Napoleone
Author-email: etienne@tomochain.com
License: GPL-3.0+
Location: /root/.local/lib/python3.6/site-packages
Requires: clint, pastel, python-slugify, docker, click
```

The next step will be to actually **START TMN**, however we cannot do this until we have two wallet addresses.
See the next section for this.

********************
TROUBLESHOOTING

FIXME extra packages might need installing
*******************
<br/>

## 7. Create Wallet Addresses
Before being able to proceed further, you will need **two** seperate Tomo addresses to operate a masternode.
One is like an ID/placeholder/dummy for the Masternode, and the other is where the 50,000 TOMO (50k) is staked from.

> * **ADDR1 - MN Address** - ID/placeholder/dummy - No coins need to be in this wallet; it can remain empty

Author: will add this

* **ADDR2 - Deposit**    - 50k Holder           - Put your staked coins here; later the 50k will go into a smart contract; rewards will show here

You will need both the **Public Key** and **Private Key** for both addresses.
It is advise that you store all of this information somewhere safe, yet accessible.
You may need to utilize this information during continued operations of your masternode.

> Password manager apps like LastPass and 1Password are your friend.

Author: Nice idea. Will add this.

Your private key is your money.
Give it to no one.

> > ADDR1 Suggestions:  

Author: Ahh. I see your confusion now. Could be a language difference thing.
I have suggested we use "ACCT1".

Many English guides call a singular wallet account an "Address". Each Wallet Address has a public key address and a private key. We use address for both and somehow don't find it confusing haha.
Ex: https://zcoin.io/zcoin-znode-setup-guide/

> > ADDR1 Suggestions:  

Reviewer: Ah yes I see ... Maybe the fact that I don't have any prior experience with blockchain technology than the technical approach I have now made me feel weird about it.
We can use `WALLET1` if you want. Might just be less prone to confusion and more accurate to use `ACCT1` :)

> > ADDR1 Suggestions:  

Reviewer: Also, just a note, an account address != an account public key.
The account address is derived from the public key but isn't the public key itself (it'a truncated version if I'm not mistaken)

> > ADDR1 Suggestions:  

Author: Changed all references to WALLET1 as discussed in https://github.com/tomochain/docs/pull/112#discussion_r249732223

Assure to use an address you dont have history on eth chain with - otherwise other will be able to see your investment history.

Because most wallets do not have Tomo as a selectable network yet, you will need to manually add the new mainnet if you have not already. 
See the first link below for the guide on how to do this.

Links for more info:
* [ADVISED: SETTING UP METAMASK, MEW, or TRUSTWALLET](http://bit.ly/2A6zrC7)
* [Using Metamask or Mobile Tomo Wallet](https://docs.tomochain.com/get-started/wallet/)
* [Links to mobile downloads and other tomo info](https://medium.com/tomochain/tomochain-all-in-one-overview-9fce16e13b5#6b8c)
* [Old Masternode guide (testnet) Section on wallets](https://medium.com/tomochain/how-to-run-a-tomochain-masternode-from-a-to-z-3793752dc3d1#0e58)
<br/>

## 8. Run TMN
Below, you will finally start your Tomochain node with a utility called `tmn`.

### Initial TMN start
**IMPORTANT:** Logout and SSH back in so that the $PATH variable takes effect.
This allows you to run `tmn` from any directory.

```shell
exit
ssh michael@178.62.127.177
```

When you first start your full node with `tmn start`, you need to give some information.

> > **--name:** The name of your full node. It should be formatted as a slug string.

Author: will mod, thx

>  
> **--pkey:** The private key of your ADDR1 wallet.
A TomoChain full node uses an account to be uniquely identified and to receive transaction fees.
Transaction fees are not rewards, and they are usually tiny.
Important note: we advise for security measures to use a fresh new account for your masternode.
This is not the account that will receive the rewards.
The rewards are sent to the account that will make the 50k TOMO initial deposit.

The command is structured like this:

```shell
tmn start --name [YOUR_NODE_NAME] --net mainnet --pkey [YOUR_ADDR1_PRIVATE_KEY]
```

We used the following command for our node (copy your own **name** & **private key**):

```shell
tmn start --name Atlantis --net mainnet --pkey cf03cb58************
```

********************
TROUBLESHOOTING

`tmn: command not found`

It might happen that your PATH is not set by default to include the default user binary directory. You can add it by adding it to your shell $PATH:

On GNU/Linux:
```shell
echo "export PATH=$PATH:$HOME/.local/bin" >> $HOME/.bashrc
exit
ssh michael@178.62.127.177
```
*******************
<br/>

## 9. Check sync status

> FIXME tmn status; inspect; top; stats.tomo website; # of blocks command; `tmn update`, `tmn --help`, etc

Author: Not yet, but it will be february if we don't get some version of this out;
Release early and often ;-)

> FIXME tmn status; inspect; top; stats.tomo website; # of blocks command; `tmn update`, `tmn --help`, etc

Reviewer: I it possible to comment it?
I guess html comment tags should work: 
```html
<!---
a comment
-->
```

> FIXME tmn status; inspect; top; stats.tomo website; # of blocks command; `tmn update`, `tmn --help`, etc

Author: Changed to "This section coming soon" and removed FIXMEs

# Tomochain Masternode Setup Guide

This is a procedural step-by-step guide to setting up your first Tomochain Masternode using the `tmn` tool.
It is meant for beginners and first-timers.

> Security Disclaimer: Despite there being mention of some security elements in this guide, there is **no implied guarentee of security**. You alone must fully secure your server. On a VPS, you will get rogue root login attempts within minutes of setup.

Reviewer: ```suggestion
Security Disclaimer:
Despite there being mention of some security elements in this guide, there is **no implied guarantee of security**.
You alone must fully secure your server.
```
One line per sentence
Also the part about ssh doesn't seem necessary to me

Author: I, and 4 others, such as Sam, who runs 8 masternodes, simply didnt know of the seriousness of defending against rogue logins. 3 weeks later, I went and looked at fail2bans stats and 13,000 attempts had occurred - mostly because I was on standard ssh port. If someone had illustrated to me the seriousness of the matter from the beginning, I would have spent the time to start. I was a bit lucky. So was Sam, who jumped on it as soon as he saw the fail2ban hits (he only had 8000 attempts), as well as others who expressed the same thing. It may not be our place, and the weak will get washed out.... but man, i wish someone had given me the above example.

## Technical Requirements / Recommendations
The following are required items and server specifications. 
[Click here for more details](https://docs.tomochain.com/masternode/requirements/)

>  * 50,000 Tomo deposit

Reviewer: ```suggestion
 * 50,000 TOMO deposit
```

 * Server (cloud-VPS or your-own)

>    * 16xCPU cores (CPU-optimized aka high-Ghz)

Reviewer: ```suggestion
   * 16 vCPU cores (Prefer higher clock speed. Usually found on "CPU optimized" cloud providers' servers)
```

   * 32GB RAM
 * Storage (Disk Space)

>    * 100+ GB for Linux operating system

Reviewer: ```suggestion
```
We don't have any recommendation for the base OS filesystem size

>    * 10GB / day of increasing data storage space (reccomend SSD-based Block Storage; low-latency, not NAS speeds)

Reviewer: Might want to reformulate this to include the base storage capacity needed for the existing chain

Author: Added wording for base storage needed

>  * 2 Tomo Addresses (see below)

Reviewer: ```suggestion
 * 2 TomoChain accounts
```

## Knowledge Requirements

>  * **VPS Setup** - You are able to setup your own cloud-hosted virtual private server (VPS), and 

Reviewer: ```suggestion
 * **VPS Setup** - You are able to setup your own cloud-hosted virtual private server (VPS)
```

   Why?
   The upkeep and troubleshooting will become more complex than this guide.
   Some commands fail and you must know what you are doing.
   
   

## Introduction

### What is a masternode?
A masternode is a computer on a decentralized blockchain network that is running 24 hours a day that keeps the system operating.
It powers the blockchain network by processing transactions and signing blocks.

### What are the benefits of a masternode?
Masternodes help support the network by creating and signing blocks, providing faster transaction times, and decentralized operations.
They utilize PoS (Proof of Stake) vs PoW (Proof of Work) consensus-building.
Masternode operators are paid a reward (coins) as an incentive for their involved investment of coin-deposit, server setup, and continued operation.

### What is a VPS?
VPS stands for Virtual Private Server.
They are paid servers hosted on a cloud-hosting-provider.
Each VPS runs an independent installation of an operating system (OS), Linux or Windows, and typically provides root access to the OS for advanced management and control.

### Why is a VPS highly recommended for masternodes?

> A VPS is recommended (and often required) for masternode setups, as you will need a dedicated static IP and 99.9% uptime to provide a stable and efficient node for the cryptocurrency’s network.

Reviewer: Not sure about the use of `cryptocurrency` here. The term might feel too limited regarding to what TomoChain is.

Author: Agreed, will change it.

Author: removed "cryptocurrency"

Unlike your home or office PC, a masternode VPS serves one purpose, to securely and efficiently run a masternode.
A VPS is online 24/7 and provides dedicated resources for the project’s decentralized network.

---
<br/>

## 1. Choose your hosting provider
Choose which VPS hosting provider you want to utilize.

The following providers are **sample** VPS providers.
You could choose elsewhere, or even your own 24/7 server.

>  * [Vultr (Beginner friendly)](https://www.vultr.com/)

Reviewer: Their block storage latency is disastrous and could almost considered as false advertising on it's performance.
I have a mail from them saying that their block storage is only supposed to be low perf storage like logs or archives... where their website clearly advertise it as being high performance.
Sorry little rant here ahaha
So regarding that, they might not be put first here

Author: VERY agreed haha. Their customer service gave me a hard time for being a power user and freaked when I told them it was a crypto network... "YOUR MINING". This ties into your prior remark that "cryptocurrency" limits you....had not thought of it that way. This is a decentralized application platform.

Author: put lower on list

>  * [DigitalOcean (Popular)](https://www.digitalocean.com/)

Reviewer: In general, I'm not sure that we should put opinionated comments like popular.
Let's just write "DigitalOcean", "Amazon AWS", "Google GCE", etc.

Author: I knew someone might say this :-) Its true though, they can get opinions elsewhere. Will fix

Author: Removed opinions

 * [AWS (Amazon)](https://aws.amazon.com/)
 * [OVH (Cheap)](https://www.ovh.com)
 * [GCE (Google)](https://cloud.google.com/compute/)
 * [Linode (Linux geek crowd)](https://www.linode.com/)

> Note on provider choice: It is encouraged for masternode operators to utilize various hosting providers so as to encourage a more decentralized network.

> We have noticed that DigitalOcean has been a very popular choice, however, if this popular provider goes down, others will get more rewards :-)

Reviewer: I'm ok with a lighter tone :P , I'll let @phamvancam2104 check that tho

Author: Can remove some of this if needed. At the very least, I will change "We have noticed" to "It has been noticed". Can remove smiley. I often write lightly and others rope me back in!

Of important note, not for this document, the masternodes seem to be centered around Asia and Singapore to a degree. I can tell because I have tested 8 different locales for servers, and the speeds and latency point toward the majority of servers being in the area of Singapore. I have seen differences of download speeds such as 250% higher vs servers in Europe or US. Latency differences as high as 1000% higher (though, that is to the stats.tomo server). 

Either way, it is an issue, and I was trying to think up easy ways to encourage people to venture out. Interested to hear @phamvancam2104 thoughts.

Author: Removed mention of DO as popular and made the wording on favorite-providers more subdued. Removed smiley.

<br/>

## 2. Start your VPS server
**Start/Boot your VPS server instance.**  
Choose **Ubuntu 18.04**.
This is an LTS version (Long Term Support).

> LTS versions are more stable and have seen less errors when installing Docker and Python.

Reviewer: If you explain LTS, it needs to be the right explanation
https://wiki.ubuntu.com/LTS

Author: Added link

You must use Ubuntu 18.04 to seek support from the wider community or Tomochain.
If you need help with this, [see this example](https://medium.com/tomochain/how-to-run-a-tomochain-masternode-from-a-to-z-3793752dc3d1#6122).

> Data Storage: It is recommended to assure that your provider has Block Storage or expandable disk space on SSD drives (more performant).
Block Storage is pay-as-you-go disk space that you can expand in the future.
You may not need it now, but you will in the future.
Some locations within a hosting provider do not have this, while others will.

> SSH-Key login: Seriously consider utilizing a **SSH-Key login** over passwords.
Some providers allow you to set it up upon server creation.
It is considerably more secure than passwords.

> Within 60 mins of your server being up, random hackers will be trying to login with guessed-passwords.

Reviewer: Once again, this is a bit specific and alarmist.
A good password will never be broken by a ssh bot. Most of the cloud providers generate strong passwords.

Author: Simplified language and removed alarmist statement

```shell
passwd
```

> Save your root password as you can get locked out of your own server if you forget it.
It is strongly recommended to use a 16+ char password with a mix of letters, numbers, special characters.

You are now logged in as root.
The root user is the administrative user in a Linux environment that has very broad privileges.
Because of the heightened privileges of the root account, you are _discouraged_ from using it on a regular basis.
This is because part of the power inherent with the root account is the ability to make very destructive changes, even by accident.

The next step is to [set up an alternative user account](https://www.digitalocean.com/community/tutorials/initial-server-setup-with-ubuntu-18-04) with a reduced scope of influence for day-to-day work.

### Create a new user
Use the following command to add a new user account that we will use to log in to from now on.
Our user is called michael, you can replace it with any username that you like.
Use the same username as your local mac/PC username and you will have an easier time logging in!

```shell

> adduser michael

Reviewer: note: `useradd` could be used instead to be more specific

Author: "useradd command wont create '/home/username' directory but adduser command will"
https://askubuntu.com/questions/345974/what-is-the-difference-between-adduser-and-useradd

```

You will be asked a few questions, starting with the account password.
Enter a strong password.
Optionally, fill in any of the additional information if you want or ignore it.
This is not required and you can just hit ENTER in any field you wish to skip.

### Grant the user administrative privileges
Now we have a new user account with regular account privileges.
However, we will need to do administrative tasks from this normal account.

To add these privileges to our new user, we need to add the new user to the **sudo** group.
This will allow our normal user to run commands with _administrative privileges_ by putting the word `sudo` before each command.

As root, run this command to add your new user to the sudo group (substitute michael with your new user):

```shell

> usermod -aG sudo michael

Reviewer: Would not be needed if use `useradd` instead of `adduser`

Author: I had no idea there even was two versions of the adduser command. I will have to then add a command to create their home directory? what would cd ~ do without a home dir? is there other reasons for useradd im missing?

Author: I tested the useradd command. No home directory and other initialization was not done. We need home directory for tmn later. Keeping as is for time being, so we can get this document out.

```

> -a stands for Append and -G is Group; sudo is the groupname you are adding to your user

### Assure the user received sudo
Lets check to make sure that the usermod command worked.

```shell

> cat /etc/group | grep sudo

Reviewer: note: the right way would be `sudo -l`

Author: Thx for new command knowledge. What is this showing me though? I read the man page and compared output and still dont get it. Im unsure these "check to make sure" commands even need to be in here in first place... cut alltogether? or u think its worth it @etienne-napoleone ?

Author: Resolving this so we can publish...

Reviewer: Sorry, missed that comment! `sudo -l` (or `sudo -l -l` for a more detailed output) let you see the current privilege that the user can gain through sudo.

```
User etienne may run the following commands on etienne-pc:

Sudoers entry:
    RunAsUsers: ALL
    RunAsGroups: ALL
    Commands:
	ALL
```

so basically, sudo lets you run command as another users. sudo is configured through the sudoers file (`/etc/sudoers`) via the command `visudo`.
The sudo configuration can be quite precise. Allowing certains users to run certain paths as certain other users.

[some examples](https://www.garron.me/en/linux/visudo-command-sudoers-file-sudo-default-editor.html)

Assure that you get a response such as `sudo:x:27:michael`.
This shows your new user (amongst other users) in the sudo group. 

To double check the sudo abilities of your new user, run this command:

```shell
groups michael
```

Assure that you get a response such as `michael : michael sudo` from the second command. 
This shows that michael is part of multiple groups, one being named the same as their username and the other being sudo.
Success!

When you eventually (not yet) log in as your new user, you can type `sudo` before commands to perform actions with superuser privileges.
Remain logged in as the root user for now, as we have more initial setup to do. 
After this, you will almost always login as your new user.
<br/>
<br/>

## 4. Configure your VPS (login as root user)
We will now prepare the [prerequisites for tmn](https://docs.tomochain.com/get-started/run-node/).
You need Python 3.6+ and Docker installed.

### Upgrade operating system

We should upgrade the Ubuntu operating system and all installed package libraries first.
You may consider doing this upgrade occasionally  in the future.
Type these two commands on the console with an ENTER after each:

```shell

> apt update

Reviewer: ```suggestion
sudo apt update
```

Author: At the top of the section, it says to be logged in as root still. The next section tells them to log in for the first time as user. Ill keep as it and resolve this, but let me know if u feel strongly it should change.

> apt upgrade

Reviewer: ```suggestion
sudo apt upgrade
```

```
> Watch out for WARNINGs or ERRORs. 
A lot of text could fly by, and you should watch all of it in case something installs incorrectly.
Google anything out of the ordinary and try to understand or fix it.

Reboot your VPS instace in case any of the upgraded components will only fully engage until rebooted fresh.

```shell
reboot
```

### Install Python

```shell

> apt install python3

Reviewer: ```suggestion
sudo apt install python3 python3-pip
```

Author: Some images of ubuntu need python3 installed explicitly

> apt install python3-pip

Reviewer: ```suggestion
```

```

Check if you have installed the right Python version (must be newer than version 3.6).

```shell
python3 --version
```

### System Security 
This topic is optional, but highly recommended.

Within a few minutes of the machine being up, bots and hackers attempt to login and probe it for weakness.
Thousands of connection-attempts can be seen within a week.
You will want to secure your machine with multiple levels of security.

At a minimum, you will want to consider:
* SSH on non-standard port
* UFW (Uncomplicated Fire Wall) (open port 30303 tcp & udp)

Even more security could consider:

> * Fail2ban

Reviewer: Should not be listed as "more security"
If ssh is already on a different port and using ssh keypairs, this package won't to much if any at all.

Author: True. Will reword.

Compare the results to what is below.

```shell
apt-key fingerprint 0EBFCD88
```

Example Results (format may be slightly different, but the string of char-numbers the same):
```
pub   rsa4096 2017-02-22 [SCEA]
      9DC8 5822 9FC7 DD38 854A  E2D8 8D81 803C 0EBF CD88
uid           [ unknown] Docker Release (CE deb) <docker@docker.com>
sub   rsa4096 2017-02-22 [S]
```

Use the following command to set up the "stable" repository.
You could get errors here if you have a release of Ubuntu that is too recent (non LSB).

```shell
sudo add-apt-repository "deb [arch=amd64] https://download.docker.com/linux/ubuntu $(lsb_release -cs) stable"
```

### Install Docker CE
Update the apt package index.
Then install the latest version of Docker CE:

```shell
sudo apt update
sudo apt install docker-ce
```

> Watch out for WARNINGs or ERRORs. 
Google anything out of the ordinary and try to understand or fix it.

Once installed, add your current user to the Docker group and verify that the user has been added.

```shell
sudo usermod -aG docker michael
groups michael
cat /etc/group | grep docker
```

Verify that Docker CE is installed correctly by running the hello-world image.

```shell

> sudo docker run hello-world

Reviewer: ```suggestion
docker run hello-world
```
No sudo, this is the whole purpose of having `michael` in the `docker` group :)

```

This command downloads a test image and runs it in a container.
When the container runs, it prints an informational message and exits.
**Docker CE is installed and running.**

Congratulations! You have already installed Python and Docker.
You have the prerequisites ready to run TomoChain’s tmn.

********************
TROUBLESHOOTING

error: could not access the docker daemon

If you have installed Docker, and get this error, you probably forgot to add your user to the docker group.
Please run this, close your session and open it again.

> `usermod -aG docker $your_user_name`

Reviewer: ```suggestion
`usermod -aG docker $(whoami)`
```
¯\\_(ツ)_/¯

> “We made a simple command line interface called tmn to easily and quickly start a TomoChain masternode.
It takes care of starting the necessary docker containers with the proper settings for you.
It will really suit you if you don’t already have a big infrastructure running.
Spin up a machine in your favorite cloud and get your masternode running in a few minutes!”

Install and update the tomochain-created `tmn` utility from pip:

```shell
pip3 install --user tmn
pip3 install -U tmn
```

> Watch out for WARNINGs or ERRORs and troubleshoot.

Check that `tmn` has been correctly installed, use the following command to show some tmn info:

```shell
pip3 show tmn

Name: tmn
Version: 0.2.5
Summary: Quickstart your masternode
Home-page: https://tomochain.com
Author: Etienne Napoleone
Author-email: etienne@tomochain.com
License: GPL-3.0+
Location: /root/.local/lib/python3.6/site-packages
Requires: clint, pastel, python-slugify, docker, click
```

The next step will be to actually **START TMN**, however we cannot do this until we have two wallet addresses.
See the next section for this.

********************
TROUBLESHOOTING

> FIXME extra packages might need installing

Reviewer: seems like this line wants to be fixed :P

*******************
<br/>

## 7. Create Wallet Addresses
Before being able to proceed further, you will need **two** seperate Tomo addresses to operate a masternode.

> One is like an ID/placeholder/dummy for the Masternode, and the other is where the 50,000 TOMO (50k) is staked from.

Reviewer: I think this part need rephrasing:
Account 1 (coinbase): is used to uniquely identify your masternode on the network and store transactions fees
Account 2 (the address who made the deposit): is like "the owner" of the masternode

Author: I agree and disagree.

I am unsure your version has the first time masternode installer in mind. It is technically accurate but I saw many confusions on chat around which account is which. I was trying to mitigate first-timer confusion by using less-loaded terminology. Though not technically-accurate, an initial understanding is what we want.

I reference ADDR1 and ADDR2 later in the document . Easily viewable and distinguishable.

I did the lack of "coinbase" because beginner users might get confused by coinbase.com. It did to me for quite some time until I saw it referenced a few times. I still am not totally sure what it refers to. Last year, I used to think: "Do I need to send coins from coinbase.com?!?" Laugh if you want, but its true. I figured it out, but it was an extra mental load I didn't need.

Also recall, their main concern at this point is: "where do I put my 50k?" The user has yet to learn how the deposit is done. It is likely unknown to them that the deposit gets sent into a separate smart contract. Referencing the 2nd account as anything other than the "Deposit account" or "Place to put your 50k", can be confusing. We later will show them how to *make* the deposit. I do see that this becomes a technically inaccurate label to future users - but they have more context to understand.

What about a mix of both:

* ACCT1 - Unique ID - is used to uniquely identify your masternode on the network and store transactions fees. No coins need to be inserted in this account; it can remain empty. Sometimes referred to as your 'coinbase' account.
* ACCT2 - 50k Deposit - your 50k of staked coins need to be placed here; later the 50k will go into a smart contract; rewards will show here

Reviewer: Yes it makes sense. Although it bugs me to use wrong terminology to fight a misunderstanding problem. It will make the problem worth in the long run.
Maybe we need an infographic that explains it in a visual way. How the account who make the deposit becomes the owner of the masternode which is identified by an account also.
Seems like something most people have a hard time wrapping their head around.
@phamvancam2104

Author: Mixing in your suggestions below @phamvancam2104 . 'Accurate terminology' depends who you ask. "Wallet" and "Address" seem to be the two most used by top masternode coins:

Dash: Wallet / account / address (in that order)
PIVX: wallet / address
Zcoin: "Deposit Address"
Trezor: Address / Account (each account can contain many addresses)

I have filled it in with the following:

* WALLET1 - Operating Wallet - used for operating the masternode, including signing blocks.
It effecitvely acts as a unique identifier of  your masternode. No coins need to be inserted in this wallet; it can remain empty.
* WALLET2 - Deposit Wallet   - your 50k of staked coins need to be placed here; later, the 50k will go into a smart contract; eventually, masternode rewards will show here

**Do not proceed if you are not confident** with the Linux command-line.
Why?
The upkeep and troubleshooting will become more complex than this guide. 
Some commands fail and you must know what you are doing.

## Requirements
Give basics; Link off to https://docs.tomochain.com/masternode/requirements/

50k Tomo; 16xCPU cores; 32GB RAM; 400 GB SSD storage (10GB / day of increasing space); 2 Tomo Addresses

## Introduction
About masternodes and why...; What? Benefits; VPS

## Start your VPS
VPS Choices; Links to examples; Links to SSH login guide

**Note:** it is recommended to assure that your provider has Block Storage or expandable disk space on SSD drives (more performant).
Block Storage is pay-as-you-go disk space that you can expand in the future.
You may not need it now, but you will in the future.
The tomo blockchain will eventually grow to be many hundreds of gigabytes and even a terrabyte.
As of Jan 2019, it is ~100 GB in size. This has nothing to do with RAM.

## Change passwords and accounts
Root; passwd; adduser; usermod; assure groups

## Configure your VPS (as root)

> Apt update, python3, ufw, fail2ban

Reviewer: Just an early comment here:
I'm not sure that we should get into security in the basic how to.
It could led to:
- "but, I followed your guide on security! Why did I get hacked?"
- "but, I followed your guide on security! Why sould I do more on that level?"

And tools like fail2ban are quite questionable.

Author: Solved this by creating more informal wiki security article and linking to it. Included generalized security text instead.
